gems/pending/openstack/openstack_event_monitor.rb

```diff
@@ -63,39 +63,32 @@ def each
     end
   end
 
-  # this private marker is really here for looks
-  # private_class_methods are marked below
-  private
+  cache_with_timeout(:event_monitor_class_cache) { Hash.new }
+  cache_with_timeout(:event_monitor_cache) { Hash.new }
+
+  def self.event_monitor_class(options)
+    key = event_monitor_key(options)
+    event_monitor_class_cache[key] ||= subclasses.detect do |event_monitor|
+      begin
+        event_monitor.available?(options)
+      rescue => e
+        $log.warn("MIQ(#{self}.#{__method__}) Error occured testing #{event_monitor} for #{options[:hostname]}. Trying other AMQP clients.  #{e.message}")
+        false
+      end
+    end || OpenstackNullEventMonitor
```

Comment:
The ||= followed by an || is really melting my head...might be clearer to read with some variables:

```ruby
event_monitor_class_cache[key] ||= begin
  detected_event_monitor = subclasses.detect do |event_monitor|
    #...
    #...stuff
    #...
  end
  detected_event_monitor || OpenstackNullEventMonitor
end
```

Thoughts?